### PR TITLE
Make error messages for `add` `edit` `delete` commands more specific

### LINKED ISSUE #244: Parse Tag and Index Bug
The bug affects 
1. Edit command where there's no tag/ unknown tag present 
2. Add command where there's no tag/ unknown tag present

For point 1, 
- Error should be invalid parameters and not invalid index.
Problem has to do with parsing the index? (take in values till a tag eg. t/ or g/ is found but here tags are not present) but the error message is incorrect

Screenshot 2022-11-02 at 1 22 41 PM

For point 2, 
- Error should be invalid tag found rather than the previous valid tag being flagged as invalid

### PR DESCRIPTION
**Pull request by jonasgwt**
Fixes #244 
* Add a check for invalid prefix
* Add a check for extraneous inputs by the user

Note: Please help to test extensively as there are many unintended ways that a user can input the commands.

### CODE REVIEW

File: src/main/java/seedu/masslinkers/logic/parser/ParserUtil.java [hunk 1]
```diff
@@ -27,7 +29,12 @@
  */
 public class ParserUtil {
 
-    public static final String MESSAGE_INVALID_INDEX = "Index is not a non-zero unsigned integer.";
+    public static final String MESSAGE_INVALID_INDEX = "Index is missing or not a non-zero unsigned integer.";
+    public static final String MESSAGE_UNEXPECTED_CHARACTERS = "There appears to be extraneous characters "
```
Comment (sltsheryl) [inline src/main/java/seedu/masslinkers/logic/parser/ParserUtil.java:33]: We could define `n/`, `g/`, `t/` as parameter prefixes in the UG under add command? 
Then, the error could phrase it as 
```Input with missing parameter prefixes found```
cuz extraneous characters may not be that clear?

File: src/main/java/seedu/masslinkers/logic/parser/AddCommandParser.java [hunk 2]
```diff
@@ -1,6 +1,7 @@
 package seedu.masslinkers.logic.parser;
```
Comment (sltsheryl) [inline src/main/java/seedu/masslinkers/logic/parser/AddCommandParser.java:1]: This bug is also affecting `mod add `and `mod delete`
`mod add t23 x`, `mod delete t23 x `gives `Index cannot be empty! Specify the index of a batchmate to be modified.` and not invalid index

addInt are deleteInt do not have this bug.

File: src/main/java/seedu/masslinkers/logic/parser/AddCommandParser.java [hunk 3]
```diff
@@ -37,8 +38,12 @@ public class AddCommandParser implements Parser<AddCommand> {
     public AddCommand parse(String args) throws ParseException {
         ArgumentMultimap argMultimap = ArgumentTokenizer.tokenize(args, PREFIX_NAME,
                 PREFIX_PHONE, PREFIX_EMAIL, PREFIX_TELEGRAM, PREFIX_GITHUB, PREFIX_INTEREST, PREFIX_MOD);
-        if (!arePrefixesPresent(argMultimap, PREFIX_NAME, PREFIX_TELEGRAM) || !argMultimap.getPreamble().isEmpty()) {
-            throw new ParseException(String.format(MESSAGE_INVALID_ARGUMENTS, AddCommand.MESSAGE_USAGE));
+        if (!argMultimap.getPreamble().isEmpty()) {
+            String invalidArgs = argMultimap.getPreamble();
+            throw new ParseException(String.format(MESSAGE_INVALID_ARGUMENTS, invalidArgs));
```
Comment (sltsheryl) [inline src/main/java/seedu/masslinkers/logic/parser/AddCommandParser.java:43]: Could add the AddCommand.MESSAGE_USAGE afterwards
Currently, doing `add ji xxsds`
gives `Invalid arguments 'ji xxsds' provided to the command!`

But we could add like what it should be after Invalid arguments...

File: src/main/java/seedu/masslinkers/logic/parser/ParserUtil.java [hunk 4]
```diff
@@ -80,6 +96,7 @@ public static Phone parsePhone(String phone) throws ParseException {
     public static Telegram parseTelegram(String handle) throws ParseException {
         requireNonNull(handle);
         String trimmedHandle = handle.trim();
+        checkForPrefix(trimmedHandle);
```
Comment (sltsheryl) [inline src/main/java/seedu/masslinkers/logic/parser/ParserUtil.java:99]: `add n/dxsds t/xsdsccdc dsxd` gives 
```
Telegram handle is invalid!
• Telegram handle must be at least 5 characters long.
• Telegram handle may only contain alphanumeric characters or underscores.
```
if we define telegram to be no space, will users say they are adding an extra parameter so the error should be ```Invalid arguments 'dsxd' instead?`

**Review by sltsheryl — COMMENTED** [review]
Tested abit so far, will continue

File: src/main/java/seedu/masslinkers/logic/parser/ParserUtil.java (hunk 1 above)
Comment (sltsheryl) [reply src/main/java/seedu/masslinkers/logic/parser/ParserUtil.java:33]: Currently:
edit 1 yoyoyooyo gives ```
There appears to be extraneous characters being supplied to the command.
Refer to help for the command format.
```
while 
add dsdxisjiji gives
`Invalid arguments 'dsdxisjiji' provided to the command!`
even though both seem to be of similar nature error.

**Comment by jonasgwt** [comment]
> Tested abit so far, will continue

okay thanks! I have made the changes

File: src/main/java/seedu/masslinkers/logic/parser/ParserUtil.java (hunk 1 above)
Comment (sltsheryl) [reply src/main/java/seedu/masslinkers/logic/parser/ParserUtil.java]: Fixed ✅

File: src/main/java/seedu/masslinkers/logic/parser/AddCommandParser.java (hunk 2 above)
Comment (sltsheryl) [reply src/main/java/seedu/masslinkers/logic/parser/AddCommandParser.java:1]: Fixed ✅

File: src/main/java/seedu/masslinkers/logic/parser/AddCommandParser.java (hunk 3 above)
Comment (sltsheryl) [reply src/main/java/seedu/masslinkers/logic/parser/AddCommandParser.java]: Fixed ✅

File: src/main/java/seedu/masslinkers/logic/parser/ParserUtil.java (hunk 4 above)
Comment (sltsheryl) [reply src/main/java/seedu/masslinkers/logic/parser/ParserUtil.java:108]: Could add a note into UG that 
Any words that come after a prefix will be taken as the parameter. 
Eg.
`edit 1 n/john smith t/john parallelisation`
`john parallelization` would be considered as the parameter for telegram handle which is invalid. 
Ie. a space does not demarcate the end of a parameter

File: src/main/java/seedu/masslinkers/logic/commands/ModAddCommand.java [hunk 5]
```diff
@@ -20,6 +20,7 @@
 public class ModAddCommand extends ModCommand {
```
Comment (sltsheryl) [inline src/main/java/seedu/masslinkers/logic/commands/ModAddCommand.java:20]: Do you think it's a bug when we do `mod find 32123212`, it accepts this?
Ie. we did not make a strict requirement that mod find can only search for mods (this requirement was enforced for mod add)

**Review by sltsheryl — COMMENTED** [review]
The rest seems fine, just these two things left

**Review by carriezhengjr — COMMENTED** [review]
There's a bug here when I enter the command `mod add c12341`.
Screenshot 2022-11-05 at 15 35 33

**Comment by jonasgwt** [comment]
**Summary of all the changes.**

* Fix error messages when the user inputs an invalid input such as `edit 1 sdnn`.
   * Originally, error message was "The student index provided is invalid."
   * Now, "Input with missing parameter prefixes found! Invalid Input: sdnn"
*  Fix error messages when the user inputs an two indexes such as `edit 1 2`.
   * Originally, error message was "The student index provided is invalid."
   * Now, "There is more than one valid index."
* Fix error messages when the user inputs an invalid input after command word such as  `edit poop 1`
   * Originally, error message was "The student index provided is invalid."
   * Now, "Input with missing parameter prefixes found! Invalid Input: poop"
* Fix error messages when user inputs an invalid prefix such as `add o/sdknn n/jodj t/scdnc` 
   * Originally, error message was "Missing arguments for the given command! "
   * Now, "The prefix o/ is not recognised by Mass Linkers."
* Fix error messages when user inputs `edit n/ksnknd 1` 
   * Originally, error message was "The student index provided is invalid."
   * Now, "Index is missing, not a non-zero unsigned integer or is not specified before valid prefixes."
* Fix error messages when user inputs `add n/sdinef t/adkn kdnekf` **(is this a bug??)**
   * Originally, error message was "Telegram handle is invalid!"
   * Now, "Telegram handle should not have any spaces."
* Fix bug where `mod mark all skdnn` is rejected and `mod mark cs2103t` is accepted.
   * Now, anything after `mod mark all` will be ignored (both examples will be accepted)
* Fix bug in `isValidModName`.
   * An error occurs when user inputs `mod add c2100`
   * Now user needs to input at least 2 alphabets before numbers for module name.


**Removed some changes that were originally committed due to feature freeze**
* Checks if user input `m/` prefix in mod commands.
* Add command usage after every error

File: src/main/java/seedu/masslinkers/logic/parser/ParserUtil.java (hunk 4 above)
Comment (jonasgwt) [reply src/main/java/seedu/masslinkers/logic/parser/ParserUtil.java:108]: Okay I've added it into the UG

File: src/main/java/seedu/masslinkers/logic/commands/ModAddCommand.java (hunk 5 above)
Comment (jonasgwt) [reply src/main/java/seedu/masslinkers/logic/commands/ModAddCommand.java:20]: hmm is this considered a bug? can we change it during the feature freeze?

**Comment by jonasgwt** [comment]
> 

okay thanks! i've fixed it in the recent commit

**Review by sltsheryl — APPROVED** [review]
LGTM thanks sm!

**Pull request merged**